File: src/services/http.service.ts
/**
 * HTTP Service - Perform HTTP/HTTPS checks
 */
import axios, { AxiosError } from 'axios';
import { NodeInfo, HTTPResult } from '../types';

export class HTTPService {
  private timeout = 10000; // 10 seconds

  async checkHTTP(url: string, node: NodeInfo): Promise<HTTPResult[][]> {
    try {
      if (!url.startsWith('http://') && !url.startsWith('https://')) {
        url = `http://${url}`;
      }

      const startTime = Date.now();
      
      try {
        const response = await axios.get(url, {
          timeout: this.timeout,
          maxRedirects: 5,
          validateStatus: () => true // Accept all status codes
        });

        const elapsed = (Date.now() - startTime) / 1000;
        const ip = response.request?.socket?.remoteAddress || 
                   response.request?.res?.socket?.remoteAddress || 
                   null;

        return [[{
          success: response.status >= 200 && response.status < 400 ? 1 : 0,
          time: Math.round(elapsed * 1000) / 1000,
          message: response.statusText || 'OK',
          statusCode: response.status.toString(),
          ip: ip || undefined
        }]];
      } catch (error: any) {
        const elapsed = (Date.now() - startTime) / 1000;
        
        if (error.code === 'ECONNABORTED' || error.message.includes('timeout')) {
          return [[{
            success: 0,
            time: Math.round(elapsed * 1000) / 1000,
            message: 'Connection timeout',
            statusCode: undefined,
            ip: undefined
          }]];
        }

        return [[{
          success: 0,
          time: Math.round(elapsed * 1000) / 1000,
          message: error.message || 'Connection error',
          statusCode: undefined,
          ip: undefined
        }]];
      }
    } catch (error: any) {
      return [[{
        success: 0,
        time: 0,
        message: error.message || 'Unknown error',
        statusCode: undefined,
        ip: undefined
      }]];
    }
  }

  async checkFromNodes(url: string, nodes: Map<string, NodeInfo>): Promise<Record<string, any[]>> {
    const tasks: Array<Promise<any>> = [];
    const nodeIds: string[] = [];

    for (const [nodeId, node] of nodes) {
      nodeIds.push(nodeId);
      tasks.push(this.checkHTTP(url, node));
    }

    const results = await Promise.all(tasks);
    const resultMap: Record<string, any[]> = {};

    for (let i = 0; i < nodeIds.length; i++) {
      const nodeId = nodeIds[i];
      const result = results[i];
      const hostname = `${nodeId}.node.check-host.net`;
      
      // Convert HTTPResult to array format
      resultMap[hostname] = result.map((httpResults: HTTPResult[]) => 
        httpResults.map((r: any) => [
          r.success,
          r.time,
          r.message,
          r.statusCode || null,
          r.ip || null
        ])
      );
    }

    return resultMap;
  }
}

File: services/http_service.py
"""
HTTP Service - Perform HTTP/HTTPS checks
"""
import aiohttp
import asyncio
from typing import List, Dict, Optional
from models.schemas import NodeInfo
from utils.helpers import parse_host_port
import time


class HTTPService:
    """Service for performing HTTP/HTTPS checks"""
    
    def __init__(self):
        self.timeout = 10  # Timeout in seconds
    
    async def check_http(self, url: str, node: NodeInfo) -> List[List]:
        """Check HTTP/HTTPS URL from a specific node"""
        try:
            # Parse URL
            if not url.startswith(("http://", "https://")):
                url = f"http://{url}"
            
            start_time = time.time()
            
            async with aiohttp.ClientSession() as session:
                try:
                    async with session.get(
                        url,
                        timeout=aiohttp.ClientTimeout(total=self.timeout),
                        allow_redirects=True
                    ) as response:
                        elapsed = time.time() - start_time
                        
                        # Get resolved IP (from connection)
                        ip = None
                        if hasattr(response.connection, 'peername'):
                            ip = response.connection.peername[0]
                        
                        return [[
                            1,  # success
                            round(elapsed, 3),
                            response.reason or "OK",
                            str(response.status),
                            ip or ""
                        ]]
                
                except asyncio.TimeoutError:
                    elapsed = time.time() - start_time
                    return [[0, round(elapsed, 3), "Connection timeout", None, None]]
                
                except aiohttp.ClientError as e:
                    elapsed = time.time() - start_time
                    error_msg = str(e)
                    return [[0, round(elapsed, 3), error_msg, None, None]]
        
        except Exception as e:
            return [[0, 0.0, str(e), None, None]]
    
    async def check_from_nodes(self, url: str, nodes: Dict[str, NodeInfo]) -> Dict[str, List[List]]:
        """Check HTTP/HTTPS from multiple nodes"""
        tasks = {
            node_id: self.check_http(url, node)
            for node_id, node in nodes.items()
        }
        
        results = await asyncio.gather(*tasks.values(), return_exceptions=True)
        
        return {
            f"{node_id}.node.check-host.net": result
            if not isinstance(result, Exception) else [[0, 0.0, str(result), None, None]]
            for node_id, result in zip(tasks.keys(), results)
        }

